src/server.rs

```diff
@@ -303,51 +393,149 @@ impl ServerState {
                     }
                 }
             }
-            Packet::UpdateReply{..} => {
-                unimplemented!();
+            Packet::UpdateReply{cookie, last_chat_seq, last_game_update_seq, last_gen} => {
+                let player_id = match self.get_player_id_by_cookie(cookie.as_str()) {
+                    Some(player_id) => player_id,
+                    None => {
+                        return Err(Box::new(io::Error::new(ErrorKind::PermissionDenied, "invalid cookie")));
+                    }
+                };
+                let opt_player = self.players.get_mut(&player_id);
+                match opt_player {
+                    Some(player) => {
+                        player.last_acked_msg_seq_num = last_chat_seq;
+                        Ok(None)
+                    }
+                    None => {
+                        return Err(Box::new(io::Error::new(ErrorKind::NotFound, "player not found")));
+                    }
+                }
             }
+
         }
-        /*
-        match action {
-            RequestAction::Connect => {
-                self.players.iter().for_each(|player| {
-                    assert_eq!(true, player.addr != addr && player.player_name != player_name);
-                });
+    }
 
-                self.players.push(Player::new(player_name, addr));
-            },
-            RequestAction::Ack                 => {},
-            RequestAction::Disconnect          => {},
-            RequestAction::JoinGame            => {},
-            RequestAction::ListPlayers         => {},
-            RequestAction::ChatMessage(String) => {},
-            RequestAction::None                => {},
+    // XXX
+    // Right now we'll be constructing all client Update packets for _every_ game slot.
+    fn construct_client_updates(&mut self) -> Result<Option<Vec<(SocketAddr, Packet)>>, Box<Error>> {
+        let mut client_updates: Vec<(SocketAddr, Packet)> = vec![];
+
+        if self.game_slots.len() == 0 {
+            return Ok(None);
+        }
+
+        // For each game slot, determine if each player has unread messages based on last_acked_msg_seq_num
+        // TODO: POOR PERFORMANCE BOUNTY
```

Comment:
Thinking about how to solve this: what if we had a vec of player IDs needing messages sent to them? This vec would be pushed to when a chat message is received, and then in `construct_client_updates` (here lol), the vec would be consumed. That would avoid iterating over all players in all game slots.

Reply:
Actually better make it a set, not a vec.

Reply:
Yeah that's a bit better, we can address this in the upcoming refactor stage.